read-fonts/src/tables/aat.rs

```diff
@@ -0,0 +1,619 @@
+        GlyphId::from(v as u16)
+    }
+}
+
+pub type LookupU16<'a> = TypedLookup<'a, u16>;
+pub type LookupU32<'a> = TypedLookup<'a, u32>;
+pub type LookupGlyphId<'a> = TypedLookup<'a, GlyphId>;
+
+/// Entry in an (extended) state table.
+pub struct StateEntry<T = ()> {
+    /// Index of the next state.
+    pub new_state: u16,
+    /// Flag values are table specific.
+    pub flags: u16,
+    /// Payload is table specific.
+    pub payload: T,
+}
+
+impl<'a, T: bytemuck::AnyBitPattern + FixedSize> FontRead<'a> for StateEntry<T> {
+    fn read(data: FontData<'a>) -> Result<Self, ReadError> {
+        let mut cursor = data.cursor();
+        let new_state = cursor.read()?;
+        let flags = cursor.read()?;
+        let remaining = cursor.remaining().ok_or(ReadError::OutOfBounds)?;
+        let payload = *remaining.read_ref_at(0)?;
+        Ok(Self {
+            new_state,
+            flags,
+            payload,
+        })
+    }
+}
+
+impl<T> FixedSize for StateEntry<T>
+where
+    T: FixedSize,
+{
+    // Two u16 fields + payload
+    const RAW_BYTE_LEN: usize = 4 + T::RAW_BYTE_LEN;
+}
+
+pub struct StateTable<'a> {
+    header: StateHeader<'a>,
+}
+
+impl<'a> StateTable<'a> {
+    /// Returns the class table entry for the given glyph identifier.
+    pub fn class(&self, glyph_id: GlyphId) -> Result<u8, ReadError> {
+        let glyph_id = glyph_id.to_u16();
+        if glyph_id == 0xFFFF {
+            return Ok(class::DELETED_GLYPH);
+        }
+        let class_table = self.header.class_table()?;
+        glyph_id
+            .checked_sub(class_table.first_glyph())
+            .and_then(|ix| class_table.class_array().get(ix as usize).copied())
+            .ok_or(ReadError::OutOfBounds)
+    }
+
+    /// Returns the entry for the given state and class.
+    pub fn entry(&self, state: u16, class: u8) -> Result<StateEntry, ReadError> {
+        // Each state has a 1-byte entry per class so state_size == n_classes
+        let n_classes = self.header.state_size() as usize;
+        let mut class = class as usize;
+        if class >= n_classes {
+            class = class::OUT_OF_BOUNDS as usize;
+        }
+        let state_array = self.header.state_array()?.data();
+        let entry_ix = state_array
+            .get(state as usize * n_classes + class)
+            .copied()
+            .ok_or(ReadError::OutOfBounds)? as usize;
+        let entry_offset = entry_ix * 4;
+        let entry_data = self
+            .header
+            .entry_table()?
+            .data()
+            .get(entry_offset..)
+            .ok_or(ReadError::OutOfBounds)?;
+        let mut entry = StateEntry::read(FontData::new(entry_data))?;
+        // For legacy state tables, the newState is a byte offset into
+        // the state array. Convert this to an index for consistency.
+        let new_state = ((entry.new_state as i32)
+            - (self.header.state_array_offset().to_u32() as i32))
+            / n_classes as i32;
```

Comment:
I'd be defensive with these operations, looks like fuzzer bait.

Reply:
Good call. Changed to `checked_sub` and added an additional check for divide by zero